src/realm/mixed.cpp

```diff
@@ -448,6 +444,99 @@ util::Optional<UUID> Mixed::get<util::Optional<UUID>>() const noexcept
     return get<UUID>();
 }
 
+static DataType get_common_type(DataType t1, DataType t2)
+{
+    // FIXME: only a few types supported
+    if (t1 == t2)
+        return t1;
+    if (t1 == type_Decimal || t2 == type_Decimal)
+        return type_Decimal;
+    if (t1 == type_Double || t2 == type_Double)
+        return type_Double;
+    if (t1 == type_Float || t2 == type_Float)
+        return type_Float;
+    return type_Int;
+}
+
+Mixed Mixed::operator+(const Mixed& rhs) const
+{
+    if (!is_null() && !rhs.is_null()) {
+        auto common_type = get_common_type(get_type(), rhs.get_type());
+        switch (common_type) {
+            case type_Int:
+                return export_to_type<Int>() + rhs.export_to_type<Int>();
+            case type_Float:
+                return export_to_type<float>() + rhs.export_to_type<float>();
+            case type_Double:
+                return export_to_type<double>() + rhs.export_to_type<double>();
+            case type_Decimal:
+                return export_to_type<Decimal128>() + rhs.export_to_type<Decimal128>();
+            default:
+                break;
+        }
+    }
+    return {};
+}
+
+Mixed Mixed::operator-(const Mixed& rhs) const
+{
+    if (!is_null() && !rhs.is_null()) {
+        auto common_type = get_common_type(get_type(), rhs.get_type());
+        switch (common_type) {
+            case type_Int:
+                return export_to_type<Int>() - rhs.export_to_type<Int>();
+            case type_Float:
+                return export_to_type<float>() - rhs.export_to_type<float>();
+            case type_Double:
+                return export_to_type<double>() - rhs.export_to_type<double>();
+            case type_Decimal:
+                return export_to_type<Decimal128>() - rhs.export_to_type<Decimal128>();
+            default:
+                break;
+        }
+    }
+    return {};
+}
+
+Mixed Mixed::operator*(const Mixed& rhs) const
+{
+    if (!is_null() && !rhs.is_null()) {
+        auto common_type = get_common_type(get_type(), rhs.get_type());
+        switch (common_type) {
+            case type_Int:
+                return export_to_type<Int>() * rhs.export_to_type<Int>();
+            case type_Float:
+                return export_to_type<float>() * rhs.export_to_type<float>();
+            case type_Double:
+                return export_to_type<double>() * rhs.export_to_type<double>();
+            case type_Decimal:
+                return export_to_type<Decimal128>() * rhs.export_to_type<Decimal128>();
+            default:
+                break;
+        }
+    }
+    return {};
+}
+
+Mixed Mixed::operator/(const Mixed& rhs) const
+{
+    if (!is_null() && !rhs.is_null()) {
+        auto common_type = get_common_type(get_type(), rhs.get_type());
+        switch (common_type) {
+            case type_Int:
+                return export_to_type<Int>() / rhs.export_to_type<Int>();
```

Comment:
how should we handle division by zero for ints (and possibly doubles)?

Reply:
In case of float, double, decimal128 infinity is returned which I think is fine. I propose that we for integers return "max". Or should we rather throw?

Reply:
I think returning the max for integers will give the expected behaviour in a query and we are free to define how this behaves here. Throwing would be too disruptive for general queries. Can you add some tests for division by zero for the numeric types? I did some digging and it looks like on some platforms, it is technically undefined to divide by zero with doubles/floats so we should at least try to catch that in our tests. We could also try to catch it at compile time using `static_assert(std::numeric_limits<double>::is_iec559)` and `static_assert(std::numeric_limits<float>::is_iec559)`

Reply:
Tests added. Static asserts added.